### ISSUE TITLE
Configure hardware from web UI

### ISSUE THREAD

**Issue by xoseperez** [xoseperez/espurna#382]
The idea is to be able to configure certain hardware using the web UI (or terminal). It could be done if phases:

* Relays
    * Type
    * GPIO
    * Logic (direct or inverted)
    * ~~Boot mode~~
    * ~~Pulse mode and time~~ 
    * Delays

* Buttons
    * Type (mode) 
    * GPIO
    * Actions
    * Delays

* LEDs
    * GPIO
    * Logic (direct or inverted)
    * Mode (already available for the first LED)

* Lights
    * Type (my92xx, dimmer, fastled)
    * Channels
    * Number of LEDs
    * GPIO
    * Logic (direct or inverted)

* Other stuff
    * I2C
    * Sensors
    * IR
    * RF
    * ...

**Comment by evolucja** [comment]
Hello,
is there any progress on GUI for hardware configuration?

I don't think there is anything more important like that. Currently managing two or more devices is a pain.. I'm completely lost in bulk of directories and firmwares for my devices.

In my enviorment there gonna be about 10 ESP8266 each of them having several buttons, relays and LEDs... Preparing an individual firmware would take hours. Updating each to new espurna would take hours too.

A skilled programmer moves in Atmos fluently but I belive for other typical users like me preparing confg files and managing it all could be a barrier not possible to pass.

I really look forward to see that feature in espurna GUI.

Good luck.

**Comment by xoseperez** [comment]
This issue must be split into sub-issues for better handling the task. A specific project has been created to track the different tasks involved in this. Check https://github.com/xoseperez/espurna/projects/2.

**Comment by zafrirron** [comment]
@xoseperez 
Wouldn't is be easier to first allow this by loading a json config file, to decouple GUI development? see  issue #1173. A great first step to allow external config.

**Comment by xoseperez** [comment]
Code in `v2` branch goes in that direction. Users will be able to upload the specific settings for the device as a JSON file or via web UI. JSON loading should (mostly) work already...

**Comment by jahonix** [comment]
Does it mean I can use something like an H801 (5-way LED driver) and add GPIs and their status change gets reported via Telnet? Currently unused GPIOs are just ignored.
The main part is: change reported via Telnet. Is that on the agenda already?

**Comment by xoseperez** [comment]
@jahonix Not 100% sure of what you mean. You can already change the status of any GPIO via telnet (`gpio 2 0` to set GPIO2 to LOW). But I guess you mean receiving push events on GPIO changes, right? Yes, configuring event sensors on the fly (for instance) will be doable in future releases.

**Comment by jahonix** [comment]
Exactly, receiving notifications for any GPIO event via Telnet would be great.
This way we can connect external push buttons/switches to an ESP8266 device (LED dimmer here) and proceed with actions in the control system (Crestron in my case, FWIW) without the need to compile an own firmware.
AFAIK this isn't possible with the current device tailored firmware (like espurna-1.13.2-huacanxing-h801.bin), right?
Thanks!

**Comment by xoseperez** [comment]
Not possible since you cannot configure the GPIOs on the fly. Yet.
But all events are logged to telnet, so if you can build the image yourself with the proper GPIO settings it should work.
